tap-version-14-specification.md

```diff
@@ -0,0 +1,659 @@
+
+If the directive starts with `# TODO`, the test is counted as a todo test,
+and the text after `TODO` is the explanation.
+
+```tap
+not ok 14 # TODO bend space and time
+```
+
+If the TODO has an explanation, it must be separated from TODO by a space.
+These tests represent a feature to be implemented or a bug to be fixed and
+act as something of an executable "things to do" list. They are not
+expected to succeed.
+
+Should a todo test point begin succeeding, the harness _may_ report it in
+some way that indicates that whatever was supposed to be done has been, and
+it should be promoted to a normal Test Point.
+
+Harnesses _must not_ treat failing `TODO` test points as a test failure.
+
+Harneses _should_ report `TODO` test points found as a list of items
+needing work, if that is appropriate for their use case.
+
+##### SKIP tests
+
+If the directive starts with `# SKIP`, the test is counted as a todo test,
+and the text after `SKIP` is the explanation.
+
+```tap
+ok 14 - mung the gums # SKIP leave gums unmunged for now
+```
+
+If the SKIP has an explanation, it must be separated from SKIP by a space.
+These tests indicate that a test was not run, or if it was, that its
+success or failure is being temporarily ignored.
+
+Harnesses _must not_ treat failing `SKIP` test points as a test failure.
+
+Harnesses _should_ report `SKIP` test points found as a list of items that
+were not tested, if that is appropriate for their use case.
+
+### YAML Diagnostics
+
+If a Test Point is followed by a 2-space indented block beginning with
+the line `  ---` and ending with the line `  ...`, separated from the Test
+Point only by comments or whitespace, then the block of lines between these
+markers will be interpreted as an inline YAML Diagnostic document.
+
+The YAML encodes a data structure that provides information about the
+preceding Test Point.
+
+For example:
+
+```tap
+not ok 3 - Resolve address
+  ---
+  message: "Failed with error 'hostname peebles.example.com not found'"
+  severity: fail
+  found:
+    hostname: 'peebles.example.com'
+    address: ~
+  wanted:
+    hostname: 'peebles.example.com'
+    address: '85.193.201.85'
+  at:
+    file: test/dns-resolve.c
+    line: 142
+  ...
+```
+
+Currently (March 2022) the data structure represented by a YAML Diagnostic
+block has not been standardized.  TAP14 Harnesses _must_ allow any data
+structures supported by their YAML parser implementation.
+
+A future version of this specification _may_ provide guidance regarding
+YAML Diagnostic fields in common usage.
+
+### Comments
+
+Lines outside of a YAML diagnostic block which begin with a `#` character
+preceeded by zero or more characters of whitespace, are comments.
+
+A Harness _may_ present these to the user, ignore them, or assign meaning
+to certain comments.
+
+A Harness _must not_ treat a test as a failure based on the presence of
+comment lines.  That is, a Harness _must_ ignore any unrecognized comment
+lines.
+
+### Pragmas
```

Comment:
I think this one is beyond the scope that was discussed, but I don't have a problem with it. It sounds useful.

Reply:
I like that idea. Since pragmas seem to be control mechanisms for harnesses that want to change the behavior/interpretation of a stream, changing the language to use _may_ provides an out for harnesses that want to ignore pragma lines entirely.

I think this would be especially true since there is no well defined set of pragmas that a harness is expected to enforce.

Reply:
I think it would be useful to be able to make a pragma as "if you don't understand this you should treat it as an error" (exclamation mark feels like the right marker for this). But that may be for the next version.

Reply:
943e7d733cbe9e4b0bf337d3cfdb35058f32242e

Reply:
943e7d733cbe9e4b0bf337d3cfdb35058f32242e